common/logger/src/struct_log.rs

```diff
@@ -396,55 +406,122 @@ impl FileStructLogThread {
     }
 }
 
-/// Sink that streams all structured logs to an address through UDP protocol
-struct UDPStructLog {
+/// Sink that streams all structured logs to an address through TCP protocol
+struct TCPStructLog {
     sender: SyncSender<StructuredLogEntry>,
 }
 
-impl UDPStructLog {
-    /// Creates new UDPStructLog and starts async thread to send results
-    pub fn start_new(udp_address: String) -> io::Result<Self> {
+impl TCPStructLog {
+    pub fn start_new(address: String) -> io::Result<Self> {
         let (sender, receiver) = mpsc::sync_channel(WRITE_CHANNEL_SIZE);
-        let socket = UdpSocket::bind("0.0.0.0:0").expect("couldn't bind to address");
-        let sink_thread = UDPStructLogThread {
-            receiver,
-            socket,
-            udp_address,
-        };
+        let sink_thread = TCPStructLogThread::new(receiver, &address);
         thread::spawn(move || sink_thread.run());
         Ok(Self { sender })
     }
 }
 
-impl StructLogSink for UDPStructLog {
+impl StructLogSink for TCPStructLog {
     fn send(&self, entry: StructuredLogEntry) {
         if let Err(e) = self.sender.try_send(entry) {
+            STRUCT_LOG_QUEUE_ERROR_COUNT.inc();
             // Use log crate macro to avoid generation of structured log in this case
             // Otherwise we will have infinite loop
             log::error!("[Logging] Failed to send structured log: {}", e);
         }
     }
 }
 
-struct UDPStructLogThread {
+struct TCPStructLogThread {
     receiver: Receiver<StructuredLogEntry>,
-    socket: UdpSocket,
-    udp_address: String,
+    address: SocketAddr,
 }
 
-impl UDPStructLogThread {
-    pub fn run(self) {
-        for entry in self.receiver {
+impl TCPStructLogThread {
+    fn new(receiver: Receiver<StructuredLogEntry>, address: &str) -> TCPStructLogThread {
+        let address = SocketAddr::from_str(address).unwrap();
```

Comment:
Does this support DNS hostnames? I'm getting this error when updating k8s to include this change:
```ERROR 2020-07-29 17:11:16 common/crash-handler/src/lib.rs:38 details = '''panicked at 'called `Result::unwrap()` on an `Err` value: AddrParseError(())', common/logger/src/struct_log.rs:441:23'''```

Reply:
It does not :(  I've updated here with resolving DNS https://github.com/libra/libra/pull/5363